lib/vrf.c

```diff
@@ -417,39 +466,74 @@ void vrf_terminate(void)
 /* Create a socket for the VRF. */
 int vrf_socket(int domain, int type, int protocol, vrf_id_t vrf_id)
 {
-	int ret = -1;
+	int ret, save_errno;
+	struct vrf *vrf = vrf_lookup_by_id(vrf_id);
 
+	vrf_switch_to_netns(vrf);
 	ret = socket(domain, type, protocol);
-
+	save_errno = errno;
+	vrf_switchback_to_initial();
+	errno = save_errno;
 	return ret;
 }
 
-/* vrf CLI commands */
-DEFUN_NOSH (vrf,
-       vrf_cmd,
-       "vrf NAME",
-       "Select a VRF to configure\n"
-       "VRF's name\n")
+int vrf_is_backend_netns(void)
+{
+	return (vrf_backend == VRF_BACKEND_NETNS);
+}
+
+void vrf_configure_backend(int vrf_backend_netns)
+{
+	vrf_backend = vrf_backend_netns;
+}
+
+int vrf_handler_create(struct vty *vty, const char *vrfname, struct vrf **vrf)
 {
-	int idx_name = 1;
-	const char *vrfname = argv[idx_name]->arg;
 	struct vrf *vrfp;
 
 	if (strlen(vrfname) > VRF_NAMSIZ) {
-		vty_out(vty,
-			"%% VRF name %s is invalid: length exceeds "
-			"%d characters\n",
-			vrfname, VRF_NAMSIZ);
+		if (vty)
+			vty_out(vty,
+			    "%% VRF name %s is invalid: length exceeds "
+			    "%d characters\n", vrfname, VRF_NAMSIZ);
+		else
+			zlog_warn("%% VRF name %s is invalid: length exceeds "
+				  "%d characters\n", vrfname, VRF_NAMSIZ);
 		return CMD_WARNING_CONFIG_FAILED;
 	}
 
 	vrfp = vrf_get(VRF_UNKNOWN, vrfname);
 
-	VTY_PUSH_CONTEXT(VRF_NODE, vrfp);
+	if (vty)
+		VTY_PUSH_CONTEXT(VRF_NODE, vrfp);
 
+	if (vrf)
+		*vrf = vrfp;
 	return CMD_SUCCESS;
 }
 
+int vrf_is_mapped_on_netns(struct vrf *vrf)
```

Comment:
that global variable is only set by zebra.
even if it is in shared code of lib/,  this is not in shared data.
I don't have a simple solution to pass this value through a zebra message , so that the other daemons can be aware of that value.
in addition to this, I have an other way of knowing that a vrf is with a netns backend or vrf lite backup.
vrf_is_mapped_on_netn() is called 5 times.
don't you think it is not better to have it factorised in a method ?

Reply:
What would be nice is to be able to use `vrf_is_backend_netns()`/`vrf_get_backed()` in the client daemons and not only in zebra, this would give us consistency as I said before. But yes, unfortunately the `ZEBRA_HELLO` message is sent from the client daemons to zebra, and not the other way round. So currently we don't have an easy method to pass information like this from zebra to the client daemons. I think we can keep `vrf_is_mapped_on_netns()` for now since it's not a big issue, but let's aim to get rid of it in the long term.

Reply:
thank you Renato